pennylane/qchem/convert.py

```diff
@@ -745,15 +657,124 @@ def _rcisd_state(cisd_solver, tol=1e-15):
     # generate all possible pairwise combinations of _single_ excitations of alpha and beta sectors
-    rowvals, colvals = np.array(list(product(c1a_configs, c1a_configs)), dtype=int).T
-    c2ab = (c2.transpose(0, 2, 1, 3).reshape(nocc * nvir, -1)).ravel()
     dict_fcimatr.update(
-        dict(zip(list(zip(rowvals, colvals)), c2ab * np.kron(c1a_signs, c1a_signs)))
+        dict(
+            zip(
+                list(product(c1a_configs, c1a_configs)),
+                np.einsum("i,j,ij->ij", c1a_signs, c1a_signs, c2ab, optimize=True).ravel(),
+            )
+        )
+    )
+
+    # filter based on tolerance cutoff
+    dict_fcimatr = {key: value for key, value in dict_fcimatr.items() if abs(value) > tol}
+
+    # convert sign parity from chemist to physicist convention (interleaving spin operators
+    # rather than commuting all spin-up operators to the left)
+    dict_fcimatr = _sign_chem_to_phys(dict_fcimatr, norb)
+
+    return dict_fcimatr
+
+
+def _ucisd_state(cisd_solver, tol=1e-15):
+    r"""Construct a wavefunction from PySCF's ``UCISD`` solver object.
+
+    The generated wavefunction is a dictionary where the keys represent a configuration, which
+    corresponds to a Slater determinant, and the values are the CI coefficients of the Slater
+    determinant. Each dictionary key is a tuple of two integers. The binary representation of these
+    integers correspond to a specific configuration: the first number represents the
+    configuration of the alpha electrons and the second number represents the configuration of the
+    beta electrons. For instance, the Hartree-Fock state :math:`|1 1 0 0 \rangle` will be
+    represented by the flipped binary string ``0011`` which is split to ``01`` and ``01`` for
+    the alpha and beta electrons. The integer corresponding to ``01`` is ``1`` and the dictionary
+    representation of the Hartree-Fock state will be ``{(1, 1): 1.0}``. The dictionary
+    representation of a state with `0.99` contribution from the Hartree-Fock state and ``0.01``
+    contribution from the doubly-excited state, i.e., :math:`|0 0 1 1 \rangle`, will be
+    ``{(1, 1): 0.99, (2, 2): 0.01}``.
+
+    Args:
+        cisd_solver (PySCF UCISD Class instance): the class object representing the UCISD calculation in PySCF
+        tol (float): the tolerance for discarding Slater determinants based on their coefficients
+
+    Returns:
+        dict: dictionary of the form ``{(int_a, int_b) :coeff}``, with integers ``int_a, int_b``
+        having binary representation corresponding to the Fock occupation vector in alpha and beta
+        spin sectors, respectively, and coeff being the CI coefficients of those configurations
+
+    **Example**
+
+    >>> from pyscf import gto, scf, ci
+    >>> mol = gto.M(atom=[['H', (0, 0, 0)], ['H', (0,0,0.71)]], basis='sto6g', symmetry='d2h')
+    >>> myhf = scf.UHF(mol).run()
+    >>> myci = ci.UCISD(myhf).run()
+    >>> wf_cisd = _ucisd_state(myci, tol=1e-1)
+    >>> print(wf_cisd)
```

Comment:
```suggestion
    >>> wf_cisd
```
We don't mention the verbose output from `pyscf`?

Reply:
There are also two PySCF lines, one corresponding to RHF and UCISD calculations, respectively. Is it worth adding these to the doc string? I don't think they have a bearing on this function.